src/hypothesis/core.py

```diff
@@ -438,6 +455,44 @@ def process_arguments_to_given(
     return arguments, kwargs, test_runner, search_strategy
 
 
+def create_new_argspec_for_given(original_argspec, generator_kwargs,
+                                 type_hint_strategies):
+    """Make an updated argspec for the wrapped test.
+
+    The new arspec excludes any arguments that were supplied to given,
+    and sets st.infer as the default value for arguments with a type-
+    hint-based strategy, so they can be omitted and the strategy
+    inserted at call time by process_arguments_to_given.
+
+    """
+    new_args = [a for a in original_argspec.args if a not in generator_kwargs]
+    new_kwonlyargs = [a for a in original_argspec.kwonlyargs
+                      if a not in generator_kwargs]
+    annots = {
+        k: v for k, v in original_argspec.annotations.items()
+        if k in (original_argspec.args + original_argspec.kwonlyargs)
+    }
+    annots['return'] = None
+
+    has_strats = [not type_hint_strategies.get(arg, st.nothing()).is_empty
+                  for arg in reversed(new_args)]
+    defaults = (st.infer,) * len(list(itertools.takewhile(bool, has_strats)))
+    if len(defaults) < sum(has_strats):
+        warnings.warn(
```

Comment:
> Agree on concrete suggestions, though it will complicate the implementation 😢

I just had in mind adding something along the lines of "e.g. by passing strategies in given or annotating every argument." to the error message. It doesn't have to be super-specific.

> I disagree on the deprecation, as we would be constraining the annotations users can use - when they may not even be type hints.
> For example, the type hint for self is obvious but would prevent all other non-type-hint annotations if this was made into an error.

This is a good point. However I think it pushes my feelings quite strongly in favour of making this not an error at all, and using the implementation where we add defaults of `not_set` to all the arguments.

It does break introspection, but I don't think it breaks introspection *more* than adding the defaults of `infer`, and it allows substantially more correct behaviour.